### PR TITLE
Test with ECC and FFDH accelerated and no bignum

### LINKED ISSUE #8005: Test with ECC and FFDH accelerated and no bignum
Add a pair of test components in `all.sh` very similar to `psa_crypto_config_accel_ecc_no_bignum` except FFDH is not disabled, and it's accelerated in the accelerated component.

Fix any issue that may arise (not really expecting any, but we don't know until we have tests).

Also update `docs/driver-only-build.md`:
- fill in the section about FFDH
- check if the section about ECC needs adjustments

**Depends on:** #7756

### PR DESCRIPTION
**Pull request by valeriosetti**
This PR:
- extends test scenario introduced by the #7754 + #7755 + #7756 series by accelerating also FFDH together with ECC.
- improves the `driver-only-builds.md` documentation as requested by the issue

Resolves #8005

## PR checklist

- [x] **changelog** deferred to #7757
- [x] **backport** not required, since it's only related to test
- [x] **tests** included - that's what the PR is about!

### CODE REVIEW

**Comment by valeriosetti** [comment]
I'm force pushing the last commit again in order to restart the CI because it looks stuck to me

File: docs/driver-only-builds.md [hunk 1]
```diff
@@ -107,6 +103,11 @@ without `MBEDTLS_ECP_C` provided the corresponding
 RSA or FFDH, then you can also disable `MBEDTLS_BIGNUM_C` for further code
 size saving.
 
+[Coming soon] As noted in the "Limitations regarding the selection of curves"
+section below, there is an upcoming requirement for all the required curves to
+be also accelerated in the PSA driver in order to exclude the builtin algs
```
Comment (tom-daubney-arm) [inline docs/driver-only-builds.md:108]:
```suggestion
section below, there is an upcoming requirement for all the required curves to
also be accelerated in the PSA driver in order to exclude the builtin algs
```

File: docs/driver-only-builds.md [hunk 2]
```diff
@@ -144,10 +145,34 @@ timeline, please let us know if you're interested.
 
 ### Limitations regarding the selection of curves
 
-TODO: apparently we don't really support having some curves built-in and
-others driver-only... investigate and describe the situation. See also #7899.
+There is an ongoing work which tries to establish a link/constrain between
+the list of supported curves and supported algorithms both in builtin and PSA
```
Comment (tom-daubney-arm) [inline docs/driver-only-builds.md:149]:
```suggestion
the list of supported curves and supported algorithms both in the builtin and PSA
```

File: docs/driver-only-builds.md [hunk 3]
```diff
@@ -144,10 +145,34 @@ timeline, please let us know if you're interested.
...
+`MBEDTLS_PSA_ACCEL` pairs:
+
+- `[PSA_WANT|MBEDTLS_PSA_ACCEL]_KEY_TYPE_DH_PUBLIC_KEY`;
+- `[PSA_WANT|MBEDTLS_PSA_ACCEL]_KEY_TYPE_DH_KEY_PAIR_BASIC`;
+- `[PSA_WANT|MBEDTLS_PSA_ACCEL]_KEY_TYPE_DH_KEY_PAIR_IMPORT`;
+- `[PSA_WANT|MBEDTLS_PSA_ACCEL]_KEY_TYPE_DH_KEY_PAIR_EXPORT`;
+- `[PSA_WANT|MBEDTLS_PSA_ACCEL]_KEY_TYPE_DH_KEY_PAIR_GENERATE`;
+
+The same holds for the associated algorithm:
+`[PSA_WANT|MBEDTLS_PSA_ACCEL]_ALG_FFDH` allow to build accelerating FFDH and
```
Comment (tom-daubney-arm) [inline docs/driver-only-builds.md:173]:
```suggestion
`[PSA_WANT|MBEDTLS_PSA_ACCEL]_ALG_FFDH` allow builds accelerating FFDH and
```

**Comment by tom-daubney-arm** [comment]
Note: The Travis build has errored out rather than it being an issue with the PR.

File: tests/scripts/all.sh [hunk 4]
```diff
@@ -2622,16 +2622,29 @@ component_test_psa_crypto_config_reference_ecc_no_ecp_at_all () {
...
+#       - accelerate all EC and FFDH algs, disable only RSA
+# - component_test_psa_crypto_config_reference_ecc_ffdh_no_bignum
+#       - this is the reference component of the above
+#       - it still disables RSA, but it uses builtin EC and FFDH algs
+#
+# This function accepts 2 parameters:
+# $1: it is a boolean values which states if we are testing an accelerated
+#     scenario or not.
+# $2: it is a string value which states which are the tested components. Allowed
+#     values are "ECC" or "ECC_DH".
```
Comment (tom-daubney-arm) [inline tests/scripts/all.sh:2644]:
```suggestion
# $1: a boolean value which states if we are testing an accelerated
#     scenario or not.
# $2: a string value which states which components are tested. Allowed
#     values are "ECC" or "ECC_DH".
```

File: tests/scripts/all.sh [hunk 5]
```diff
@@ -2666,29 +2679,56 @@ config_psa_crypto_config_accel_ecc_no_bignum() {
...
-# Keep in sync with component_test_psa_crypto_config_reference_ecc_no_bignum()
-component_test_psa_crypto_config_accel_ecc_no_bignum () {
-    msg "build: full + accelerated EC algs + USE_PSA - ECP - BIGNUM"
+# The goal is to build and test accelerating either:
+# - ECC only or
+# - both ECC and FFDH
+#
+# It is meant to be used in conjunction with
+# common_test_psa_crypto_config_reference_ecc_ffdh_no_bignum() for drivers'
+# coverage analysis in "analyze_outcomes.py" script.
```
Comment (tom-daubney-arm) [inline tests/scripts/all.sh:2716]:
```suggestion
# It is meant to be used in conjunction with
# common_test_psa_crypto_config_reference_ecc_ffdh_no_bignum() for driver
# coverage analysis in the "analyze_outcomes.py" script.
```

File: tests/scripts/all.sh [hunk 6]
```diff
@@ -2666,29 +2679,56 @@ config_psa_crypto_config_accel_ecc_no_bignum() {
...
+common_test_psa_crypto_config_accel_ecc_ffdh_no_bignum () {
+    TEST_TARGET="$1"
+
+    # This is an internal helper to simplify text messages' handling
```
Comment (tom-daubney-arm) [inline tests/scripts/all.sh:2720]:
```suggestion
    # This is an internal helper to simplify text message handling
```

File: tests/scripts/all.sh [hunk 7]
```diff
@@ -2719,41 +2769,73 @@ component_test_psa_crypto_config_accel_ecc_no_bignum () {
...
+# components) compared to the ones used in
+# common_test_psa_crypto_config_accel_ecc_ffdh_no_bignum() above.
+#
+# It is meant to be used in conjunction with
+# common_test_psa_crypto_config_accel_ecc_ffdh_no_bignum() for drivers'
+# coverage analysis in "analyze_outcomes.py" script.
+common_test_psa_crypto_config_reference_ecc_ffdh_no_bignum () {
+    TEST_TARGET="$1"
+
+    # This is an internal helper to simplify text messages' handling
```
Comment (tom-daubney-arm) [inline tests/scripts/all.sh:2803]:
```suggestion
    # This is an internal helper to simplify text message handling
```

**Review by tom-daubney-arm — APPROVED** [review]
Thank you for making the various edits.

LGTM now.

**Comment by mpg** [comment]
@paul-elliott-arm How's your review bandwidth looking? This is a pre-requisite for #7757 which is a MUST for the release (code-freeze end of next week as you know), so we should try to have this merged by the end of this week. If you don't have enough review bandwidth, please let us know, so that Tom C or I can take over and review this (I just came back and driver-only ECC is my top priority until everything we need in the release is done, so I have the bandwidth this week).

**Review by paul-elliott-arm — CHANGES_REQUESTED** [review]

**Comment by paul-elliott-arm** [comment]
I am adopting this as Valerio is on leave. I have made the one change I myself requested, @tom-daubney-arm, if you could re-review, and @tom-cosgrove-arm if you could take second review, that would be much appreciated.

File: docs/driver-only-builds.md [hunk 8]
```diff
@@ -145,9 +145,9 @@ timeline, please let us know if you're interested.
 
 ### Limitations regarding the selection of curves
 
-There is an ongoing work which tries to establish a link/constrain between
-the list of supported curves and supported algorithms both in the builtin and
-PSA sides. In particular:
+There is ongoing work which is trying to establish the links and constraints
```
Comment (tom-cosgrove-arm) [inline docs/driver-only-builds.md:148]: You could just say "There is ongoing work to establish the links and constraints..."

Comment (tom-daubney-arm) [reply docs/driver-only-builds.md:148]: Agree, prefer this wording

Comment (mpg) [reply docs/driver-only-builds.md:148]: I'll point out this sentence is going to be changed again in #7757 once this PR and #8075 are in, so I don't think it's essential to get the wording perfect right now (we can be much more picky in the PR the implements #7757).

**Review by tom-daubney-arm — APPROVED** [review]
Given that we are not too fussed about that wording nit right now, this PR LGTM.

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Pull request merged**